android/app/src/main/java/com/msupplymobile/Ble/BleManager.java

```diff
@@ -0,0 +1,192 @@
+    private void returnException(MsupplyException exception){
+        if (Debug.LOG) Log.i(Debug.TAG, "BleManager: Returning Exception");
+        WritableMap jsObject = Arguments.createMap();
+        jsObject.putBoolean("success", false);
+        jsObject.putMap("error", exception.toObject());
+        promise.resolve(jsObject);
+    }
+
+    /**
+     * Exit point for returning a result to JS.
+     * Returns all succesful executions to JS with
+     * an object in the shape:
+     * {
+     *     success: true,
+     *     data: [ BleDevice.toObject(), .. ] 
+     * }
+     */
+    private void returnResult(){
+        if (Debug.LOG) Log.i(Debug.TAG, "Return Result");
+        WritableArray results = Arguments.createArray();
+        
+        for (String key : this.devices.keySet()){
+            try{
+                if (Debug.LOG) Log.i(Debug.TAG, key);
+                WritableMap bleDeviceAsObject = ((BleDevice)this.devices.get(key)).toObject();
+                results.pushMap(bleDeviceAsObject);
+            } catch(Exception exception){
+                returnException(new MsupplyException(ErrorCode.E_PARSING_FAILED, exception));
+                return;
+            }
+        }
+        WritableMap jsObject = Arguments.createMap();
+        jsObject.putBoolean("success", true);
+        jsObject.putArray("data", results);
+        promise.resolve(jsObject);
+    }
+    
+    /**
+     * Spin up a BLE device scanner and register to it as a listener.
+     * If providing a device address, scan for that specific device,
+     * otherwise if providing an empty string scan for all devices.
+     * On either fetching the advertisement data for one device, all the
+     * time expires when scanning for all devices, the scanner will notify
+     * of all fetched advertisement data.
+     */
+    @ReactMethod
+    public void getDevices(int manufacturerID, String macAddress, Promise promise) {
+        if (Debug.LOG) Log.i(Debug.TAG, "getDevices:" + Integer.toString(manufacturerID) + " " + macAddress);
+        this.promise = promise;
+        BleDeviceScanner deviceScanner = new BleDeviceScanner(reactContext);
+        deviceScanner.registerListener(this);
+        deviceScanner.setFilters(manufacturerID, macAddress);
+        // Full coverage fallback catcher for any uncaught errors occuring.
+        try{
+            deviceScanner.startScan();
+        }catch(Throwable e){
+            promise.reject("error", e.toString());
+        }
+    }
+
+    /**
+     * Find the provided device and send the provided command.
+     */
+    @ReactMethod
+    public void sendCommand(int manufacturerID, String macAddress, String command, Promise promise) {
+        if (Debug.LOG) Log.i(Debug.TAG, "sendCommand" + Integer.toString(manufacturerID) + " " + macAddress + " " + command);
+        this.promise = promise;
+        this.command = command;
+        this.macAddress = macAddress;
+        getDevices(manufacturerID, macAddress, promise);
```

Comment:
20a99ca1c79ec1459a804d7e29fb89ec86f3275c

This was the main reason for the change. As react-native native classes are singletons I thought it was best it didn't keep old state as it can lead to pretty hard to find bugs in the future